builtin/credential/approle/path_role.go

```diff
@@ -0,0 +1,1544 @@
+package approle
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/fatih/structs"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/helper/policyutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+// roleStorageEntry stores all the options that are set on an role
+type roleStorageEntry struct {
+	// UUID that uniquely represents this role. This serves as a credential
+	// to perform login using this role.
+	RoleID string `json:"role_id" structs:"role_id" mapstructure:"role_id"`
+
+	// UUID that serves as the HMAC key for the hashing the 'secret_id's
+	// of the role
+	HMACKey string `json:"hmac_key" structs:"hmac_key" mapstructure:"hmac_key"`
+
+	// Policies that are to be required by the token to access this role
+	Policies []string `json:"policies" structs:"policies" mapstructure:"policies"`
+
+	// Number of times the SecretID generated against this role can be
+	// used to perform login operation
+	SecretIDNumUses int `json:"secret_id_num_uses" structs:"secret_id_num_uses" mapstructure:"secret_id_num_uses"`
+
+	// Duration (less than the backend mount's max TTL) after which a
+	// SecretID generated against the role will expire
+	SecretIDTTL time.Duration `json:"secret_id_ttl" structs:"secret_id_ttl" mapstructure:"secret_id_ttl"`
+
+	// Duration before which an issued token must be renewed
+	TokenTTL time.Duration `json:"token_ttl" structs:"token_ttl" mapstructure:"token_ttl"`
+
+	// Duration after which an issued token should not be allowed to be renewed
+	TokenMaxTTL time.Duration `json:"token_max_ttl" structs:"token_max_ttl" mapstructure:"token_max_ttl"`
+
+	// A constraint, if set, requires 'secret_id' credential to be presented during login
```

Comment:
Note per our Slack discussion - we shouldn't have any required bindings, so it should be valid to turn this off and only use CIDR or only use the role ID.